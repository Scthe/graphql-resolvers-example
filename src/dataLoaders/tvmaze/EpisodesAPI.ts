import { IDelegateToSchemaOptions } from "apollo-server-express";
import objectHash from "object-hash";
import DataLoader from "dataloader";

import RestResource, {
  DataloaderReturnType,
  ResourceList,
} from "../RestResource";
import * as tvmaze from "./tvmaze.api";

/*** Structure of the object returned from tvmaze API */
export interface Episode {
  id: ID;
  idx: number;
  name: string;
  summary?: string;
  airdate: string;
  runtime: number;
}
/** Type returned from listing endpoint */
// type EpisodeListItem = Episode;

type EpisodeBySeasonKey = { showId: ID; seasonId: ID };

export default class EpisodesAPI extends RestResource {
  constructor() {
    super(tvmaze.URL);
  }

  getOne = async (id: ID): Promise<Episode> => this.dataLoader.load(id);

  findBySeason = async (showId: ID, seasonId: ID): ResourceList =>
    this.bySeasonDataLoader.load({
      showId,
      seasonId,
    });

  private addToCache = (...item: Episode[]): void => {
    item.forEach((item) => this.dataLoader.prime(item.id, item));
  };

  private getByIds = async (
    ids: readonly ID[]
  ): DataloaderReturnType<Episode> => {
    const promises = ids.map((id) => this.get<Episode>(`/episodes/${id}`));
    const result = await Promise.allSettled(promises);
    return this.collectSettledPromises(result);
  };

  private search = async (
    keys: readonly EpisodeBySeasonKey[]
  ): DataloaderReturnType<ID[]> => {
    const promises = keys.map((key) =>
      this.get<Episode[]>(`seasons/${key.seasonId}/episodes`)
    );
    const result = await Promise.allSettled(promises);

    return this.collectSettledPromises(result, (episodes) => {
      this.addToCache(...episodes);
      return episodes.map((e) => e.id);
    });
  };

  // has to be on the end of the file, cause `this.getByIds` is undefined otherwise
  private dataLoader = new DataLoader(this.getByIds);

  // Separate dataloader that maps {showId, seasonId} to ids of episodes.
  private bySeasonDataLoader = new DataLoader(this.search, {
    // IMPORTANT: this dataloader has a composite key. We could have used just seasonID,
    // but this is a better showcase od DataLoader.
    cacheKeyFn: (key: EpisodeBySeasonKey) => objectHash(key),
  });
}
