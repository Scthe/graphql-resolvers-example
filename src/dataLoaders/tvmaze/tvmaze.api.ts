export const URL = "https://api.tvmaze.com";

export type ApiListResponse<Key extends string | number | symbol, T extends object> = {
  [k in Key]: T;
} & {
  score: number;
}
