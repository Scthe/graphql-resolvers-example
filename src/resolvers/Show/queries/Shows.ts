import { QueryShowsArgs } from "typingsGql";
import { RootType as ShowsListRoot } from "../types/ShowsList";

export default (_root: any, args: QueryShowsArgs): ShowsListRoot => ({
  name: args.name || "",
});
