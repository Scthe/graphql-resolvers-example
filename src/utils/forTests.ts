import nock from "nock";

// utils used in tests

export const setupNock = (): void => {
  beforeEach(() => {
    // each jest test loads modules anew
    nock.disableNetConnect();
    nock.enableNetConnect("127.0.0.1");
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    nock.restore(); // https://github.com/nock/nock/issues/1817
  });
};
