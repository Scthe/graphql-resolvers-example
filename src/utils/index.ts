// eslint-disable @typescript-eslint/no-explicit-any
export const parseId = (id: number | string): ID => {
  if (typeof id === "string") {
    return parseInt(id, 10) as any;
  }
  return id as any;
};
