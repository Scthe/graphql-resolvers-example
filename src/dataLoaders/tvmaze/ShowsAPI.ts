import DataLoader from "dataloader";

import RestResource, {
  DataloaderReturnType,
  ResourceList,
} from "../RestResource";
import * as tvmaze from "./tvmaze.api";

/*** Structure of the object returned from tvmaze API */
export interface Show {
  id: ID;
  name: string;
  type: string;
  language: string;
  status: string;
  runtime: number;
  premiered: string;
  genres: string[];
  network: string;
  country: string;
  summary: string;
}

type ShowsListItem = tvmaze.ApiListResponse<"show", Show>;

export default class ShowsAPI extends RestResource {
  constructor() {
    super(tvmaze.URL);
  }

  getOne = async (id: ID): Promise<Show> => this.dataLoader.load(id);

  findByName = async (name: string): ResourceList =>
    this.searchDataLoader.load(name);

  private addToCache = (item: Show): void => {
    this.dataLoader.prime(item.id, item);
  };

  private getByIds = async (ids: readonly ID[]): DataloaderReturnType<Show> => {
    // We get array of ids, normally we would want to use api like:
    // `shows/?id=1&id=2&id=42` to get all data in one request (thus solving n+1).
    // Since our API does not have this, we need to do this manually.
    const promises = ids.map((id) => this.get<Show>(`shows/${id}`));
    const result = await Promise.allSettled(promises);

    // It is important that we return objects in same order as `ids` were provided.
    // Since we do not change the order, we can just return result.
    return this.collectSettledPromises(result);
  };

  private search = async (
    searchNames: readonly string[]
  ): DataloaderReturnType<ID[]> => {
    const name = searchNames[0]; // batching is off for this request
    const items = await this.get<ShowsListItem[]>(`search/shows?q=${name}`);
    const ids = items.map((e) => {
      this.addToCache(e.show);
      return e.show.id;
    });
    return [ids]; // wrap in array cause `searchNames` is an array
  };

  // has to be on the end of the file, cause `this.getByIds` is undefined otherwise
  private dataLoader = new DataLoader(this.getByIds);

  // Separate dataloader that maps search phrase to ID[].
  // Since ID[] is not the same as `Show`.
  // We use dataloader, since there is no separate function to get totalCount of shows.
  // In our resolver implementation we call `findByName` twice (once for data and once for totalCount), and we want to deduplicate that with cache.
  private searchDataLoader = new DataLoader(this.search, {
    batch: false, // turn off batching - we will always get a single item to `this.search`
  });
}
