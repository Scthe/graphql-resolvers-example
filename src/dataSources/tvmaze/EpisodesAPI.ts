import RestResource, { ResourceList } from "../RestResource";
import MyDataLoader, { DataLoaderReturnType } from "../MyDataLoader";
import * as tvmaze from "./tvmaze.api";

/*** Structure of the object returned from tvmaze API */
export interface Episode {
  id: ID;
  idx: number;
  name: string;
  summary?: string;
  airdate: string;
  runtime: number;
}

type EpisodeBySeasonKey = { showId: ID; seasonId: ID };

/** Api to get episodes data */
export default class EpisodesAPI extends RestResource {
  constructor() {
    super(tvmaze.URL);
  }

  getOne = async (id: ID): Promise<Episode> => this.dataLoader.load(id);

  findBySeason = async (showId: ID, seasonId: ID): ResourceList =>
    this.bySeasonDataLoader.load({
      showId,
      seasonId,
    });

  private addToCache = (...item: Episode[]): void => {
    item.forEach((item) => this.dataLoader.prime(item.id, item));
  };

  private getByIds = async (
    ids: readonly ID[]
  ): DataLoaderReturnType<Episode> => {
    this.debugLog("getByIds", ids);

    const promises = ids.map((id) => this.get<Episode>(`/episodes/${id}`));
    const result = await Promise.allSettled(promises);
    return MyDataLoader.collectSettledPromises(result);
  };

  private getEpisodesForSeasons = async (
    keys: readonly EpisodeBySeasonKey[]
  ): DataLoaderReturnType<ID[]> => {
    this.debugLog("getEpisodesForSeasons", keys);

    const promises = keys.map((key) =>
      this.get<Episode[]>(`seasons/${key.seasonId}/episodes`)
    );
    const result = await Promise.allSettled(promises);

    return MyDataLoader.collectSettledPromises(result, (episodes) => {
      this.addToCache(...episodes);
      return episodes.map((e) => e.id);
    });
  };

  // Map<EpisodeId, Episode>
  private dataLoader = new MyDataLoader(this.getByIds);

  // Map<{ShowId, SeasonId}, EpisodeId[]>
  //
  // IMPORTANT: this dataloader has a composite key. We could have used just seasonID,
  // but this is a better showcase od DataLoader.
  private bySeasonDataLoader = new MyDataLoader(this.getEpisodesForSeasons);
}
