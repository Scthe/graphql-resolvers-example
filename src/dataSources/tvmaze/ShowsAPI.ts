import RestResource, { ResourceList } from "../RestResource";
import MyDataLoader, { DataLoaderReturnType } from "../MyDataLoader";
import * as tvmaze from "./tvmaze.api";

/*** Structure of the object returned from tvmaze API */
export interface Show {
  id: ID;
  name: string;
  type: string;
  language: string;
  status: string;
  runtime: number;
  premiered: string;
  genres: string[];
  summary: string;
  externals: {
    imdb: string;
  };
}

export type ShowSearchItem = tvmaze.SearchResponseItem<"show", Show>;

/** Api to get shows data */
export default class ShowsAPI extends RestResource {
  constructor() {
    super(tvmaze.URL);
  }

  getOne = async (id: ID): Promise<Show> => this.dataLoader.load(id);

  getByImdbId = async (imdbId: string): Promise<ID> => {
    // TBH we could cache result of this request in `Map<imdbId, ID>`,
    // but no need to complicate for now.
    const show = await this.get<Show>(`/lookup/shows?imdb=${imdbId}`);
    this.addToCache(show);
    return show.id;
  };

  findByName = async (name: string): ResourceList =>
    this.searchDataLoader.load(name);

  private addToCache = (item: Show): void => {
    this.dataLoader.addToCache("id", item);
  };

  private getByIds = async (ids: readonly ID[]): DataLoaderReturnType<Show> => {
    this.debugLog("getByIds", ids);
    // We get array of ids, normally we would want to use api like:
    // `shows/?id=1&id=2&id=42` to get all data in one request (thus solving n+1).
    // Since our API does not have this, we need to do this manually.
    const promises = ids.map((id) => this.get<Show>(`shows/${id}`));
    const result = await Promise.allSettled(promises);

    // It is important that we return objects in same order as `ids` were provided.
    // Since we do not change the order, we can just return result.
    return MyDataLoader.collectSettledPromises(result);
  };

  private searchByShowNames = async (
    searchNames: readonly string[]
  ): DataLoaderReturnType<ID[]> => {
    this.debugLog("searchByShowNames", searchNames);

    const name = searchNames[0]; // batching is off for this request
    const items = await this.get<ShowSearchItem[]>(`search/shows?q=${name}`);
    const ids = items.map((e) => {
      this.addToCache(e.show);
      return e.show.id;
    });
    return [ids]; // wrap in array cause `searchNames` is an array
  };

  // Map<ShowId, Show>
  private dataLoader = new MyDataLoader(this.getByIds);

  // Map<searchedShow, ShowId[]>
  // Separate dataloader that maps search phrase to ID[].
  //
  // We use dataloader, since there is no separate function to get totalCount of shows.
  // In our resolver implementation we call `findByName` twice (once for data and once for totalCount), and we want to deduplicate that with cache.
  private searchDataLoader = new MyDataLoader(this.searchByShowNames, {
    batch: false, // turn off batching - we will always get a single item to `this.search`
  });
}
