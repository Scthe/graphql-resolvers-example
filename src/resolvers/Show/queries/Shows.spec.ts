import supertest from "supertest";
import nock, { Scope } from "nock";

import app from "../../../index";
import * as gqlExamples from "./Shows.example";
import { ShowsList as GqlSchemaShowsList } from "typingsGql";
import { setupNock } from "utils/forTests";
import { parseId } from "utils";

import ShowMock from "dataSources/tvmaze/mocks/show.mock.json";
import { ShowSearchItem } from "dataSources/tvmaze/ShowsAPI";

describe("GraphQL: shows()", () => {
  let scope: Scope | null = null;

  setupNock();

  afterEach(() => {
    scope && scope.done(); // verify all planned requests were actually done
  });

  const mockSuccessShowsResponse = () => {
    // NOTE: this allows for ONE request
    const createOneResult = (score: number, id: number): ShowSearchItem => ({
      score,
      show: {
        ...ShowMock,
        id: parseId(id),
        name: `show-${id}`,
      },
    });
    scope = nock("https://api.tvmaze.com/")
      .get("/search/shows?q=House")
      .reply(200, [createOneResult(1, 1), createOneResult(5, 999)]);
  };

  test("should fetch shows list", () => {
    mockSuccessShowsResponse();

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_SHOWS,
      })
      .then((resp) => {
        const result = resp.body.data.shows as GqlSchemaShowsList;
        expect(result.meta.totalCount).toBe(2);
        expect(result.node[0].id).toBe("1");
        expect(result.node[1].id).toBe("999");
        expect(result).toMatchSnapshot(); // just for a good measure
      });
  });
});
