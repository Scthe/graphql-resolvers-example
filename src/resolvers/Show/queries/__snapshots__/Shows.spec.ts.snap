// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL: shows() should fetch shows list 1`] = `
Object {
  "meta": Object {
    "totalCount": 2,
  },
  "node": Array [
    Object {
      "id": "1",
      "name": "show-1",
    },
    Object {
      "id": "999",
      "name": "show-999",
    },
  ],
}
`;
